x/config/chooser.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"go.uber.org/yarpc/api/peer"
+	peerbind "go.uber.org/yarpc/peer"
+)
+
+// PeerListConfig facilitates decoding and building peer choosers.
+// The peer chooser combines a peer list (for the peer selection strategy, like
+// least-pending or round-robin) with a peer list binder (like static peers or
+// dynamic peers from DNS or watching a file in a particular format).
+type PeerListConfig struct {
+	Peer string       `config:"peer,interpolate"`
+	Etc  attributeMap `config:",squash"`
+}
+
+// BuildChooser translates a chooser configuration into a peer chooser, backed
+// by a peer list bound to a peer list binder.
+func (c PeerListConfig) BuildChooser(transport peer.Transport, identify func(string) peer.Identifier, kit *Kit) (peer.Chooser, error) {
+	// Establish a peer selection strategy.
+
+	// Special case for single-peer outbounds.
+	if c.Peer != "" {
+
+		if len(c.Etc) > 0 {
+			return nil, fmt.Errorf(`unrecognized attributes in peer list config: %+v`, c.Etc)
+		}
+
+		return peerbind.NewSingle(identify(c.Peer), transport), nil
+	}
+
+	// All multi-peer choosers may combine a peer list (for sharding or
+	// load-balancing) and a peer list updater.
+
+	// Find a property name that corresponds to a peer chooser/list and construct it.
+	for peerListName := range c.Etc {
+		peerListSpec := kit.peerListSpec(peerListName)
+		if peerListSpec == nil {
+			continue
```

Comment:
There are separate specs for peer list and peer list updater, and Decode does not destructively disassemble the input map, so there's no way to validate against extraneous attributes if we collapse them into a single namespace.